src/Workspaces/Core/Portable/EmbeddedLanguages/Json/JsonParser.cs

```diff
@@ -0,0 +1,476 @@
+            {
+                return null;
+            }
+        }
+
+        private JsonTree ParseTree(bool strict)
+        {
+            var arraySequence = this.ParseSequence();
+            Debug.Assert(_lexer.Position == _lexer.Text.Length);
+            Debug.Assert(_currentToken.Kind == JsonKind.EndOfFile);
+
+            var root = new JsonCompilationUnit(arraySequence, _currentToken);
+
+            var diagnostic = GetDiagnostic(root) ?? CheckTopLevel(_lexer.Text, root);
+            if (diagnostic == null)
+            {
+                diagnostic = strict
+                    ? new StrictSyntaxChecker().CheckSyntax(root)
+                    : new JsonNetSyntaxChecker().CheckSyntax(root);
+            }
+
+            var diagnostics = diagnostic == null
+                ? ImmutableArray<EmbeddedDiagnostic>.Empty
+                : ImmutableArray.Create(diagnostic.Value);
+
+            return new JsonTree(
+                _lexer.Text, root, diagnostics);
+        }
+
+        private EmbeddedDiagnostic? CheckTopLevel(
+            ImmutableArray<VirtualChar> text, JsonCompilationUnit compilationUnit)
+        {
+            var arraySequence = compilationUnit.Sequence;
+            if (arraySequence.ChildCount == 0)
+            {
+                if (text.Length > 0 &&
+                    compilationUnit.EndOfFileToken.LeadingTrivia.All(
+                        t => t.Kind == JsonKind.WhitespaceTrivia || t.Kind == JsonKind.EndOfLineTrivia))
+                {
+                    return new EmbeddedDiagnostic(WorkspacesResources.Syntax_error, GetSpan(text));
+                }
+            }
+            else if (arraySequence.ChildCount >= 2)
+            {
+                var firstToken = GetFirstToken(arraySequence.ChildAt(1).Node);
+                return new EmbeddedDiagnostic(
+                    string.Format(WorkspacesResources._0_unexpected, firstToken.VirtualChars[0].Char),
+                    firstToken.GetSpan());
+            }
+            foreach (var child in compilationUnit.Sequence)
+            {
+                if (child.IsNode && child.Node.Kind == JsonKind.EmptyValue)
+                {
+                    var emptyValue = (JsonEmptyValueNode)child.Node;
+                    return new EmbeddedDiagnostic(
+                        string.Format(WorkspacesResources._0_unexpected, ','),
+                        emptyValue.CommaToken.GetSpan());
+                }
+            }
+
+            return null;
+        }
+
+        private static JsonToken GetFirstToken(JsonNode node)
+        {
+            foreach (var child in node)
+            {
+                return child.IsNode
+                    ? GetFirstToken(child.Node)
+                    : child.Token;
+            }
+
+            throw new InvalidOperationException();
+        }
+
+        private static EmbeddedDiagnostic? GetDiagnostic(JsonNode node)
```

Comment:
A few of the overloads could be renamed `GetFirstDiagnostic`.

Reply:
done.  #resolved